tiledb/sm/cpp_api/arrow_io_impl.h

```diff
@@ -0,0 +1,834 @@
+   * the release callback of the source is set to null.
+   */
+  CPPArrowArray(
+      int64_t elem_num,
+      int64_t null_num,
+      int64_t offset,
+      std::vector<std::shared_ptr<CPPArrowArray>> children,
+      std::shared_ptr<CPPArrowArray> dictionary,
+      std::vector<void*> buffers)
+      : array_((ArrowArray*)malloc(sizeof(ArrowArray))) {
+    (void)dictionary;  // not supported by TileDB
+
+    array_ = (ArrowArray*)malloc(sizeof(ArrowArray));
+    if (array_ == nullptr)
+      throw tiledb::TileDBError("Failed to allocate ArrowArray");
+
+    // Data description
+    array_->length = elem_num;
+    array_->null_count = null_num;
+    array_->offset = offset;
+    array_->n_buffers = (int64_t)buffers.size();
+    array_->n_children = (int64_t)children.size();
+    array_->buffers = nullptr;
+    array_->children = nullptr;
+    array_->dictionary = nullptr;
+    // Bookkeeping
+    array_->release = ([](ArrowArray* array_p) {
+      assert(array_p->release != nullptr);
+
+      // Release children
+      for (int64_t i = 0; i < array_p->n_children; i++) {
+        ArrowArray* child_array = array_p->children[i];
+        child_array->release(child_array);
+        assert(child_array->release == nullptr);
+      }
+
+      // Release dictionary
+      struct ArrowArray* dict = array_p->dictionary;
+      if (dict != NULL && dict->release != NULL) {
+        dict->release(dict);
+        assert(dict->release == NULL);
+      }
+
+      // mark the ArrowArray struct as released
+      array_p->release = nullptr;
+
+      delete (CPPArrowArray*)(array_p->private_data);
+    });
+    array_->private_data = this;
+
+    buffers_ = buffers;
+    array_->buffers = (const void**)buffers_.data();
+  }
+
+  /*
+   * CPPArrowArray destructor
+   *
+   * This destructor is invoked via the ArrowArray.release
+   * callback. Owned member data is released via default destructors.
+   */
+  ~CPPArrowArray() = default;
+
+  void export_ptr(ArrowArray* array) {
+    memcpy(array, array_, sizeof(ArrowArray));
+    free(array_);
+    array_ = nullptr;
+  }
+
+  ArrowArray* ptr() const {
+    assert(array_ != nullptr);
+    return array_;
+  }
+
+  ArrowArray* mutable_ptr() {
+    assert(array_ != nullptr);
+    return array_;
+  }
+
+ private:
+  ArrowArray* array_;
+  std::vector<void*> buffers_;
+};
+
+/* ****************************** */
+/*         Arrow Importer         */
+/* ****************************** */
+
+class ArrowImporter {
+ public:
+  ArrowImporter(std::shared_ptr<tiledb::Query> query);
+  ~ArrowImporter();
+
+  void import_(std::string name, ArrowArray* array, ArrowSchema* schema);
+
+ private:
+  std::shared_ptr<tiledb::Query> query_;
+  std::vector<void*> offset_buffers_;  //
+
+};  // class ArrowExporter
+
+ArrowImporter::ArrowImporter(std::shared_ptr<tiledb::Query> query) {
+  query_ = query;
+}
+
+ArrowImporter::~ArrowImporter() {
+  for (auto p : offset_buffers_) {
+    free(p);
```

Comment:
We should probably use `std::free` and `std::malloc` for the STL verison

Reply:
👍